Notebooks/Table Header Detection.ipynb

```diff
@@ -567,44 +582,124 @@
+    "    similarity[feature + 'A' + suffix] = cell[feature] and neighbour[feature]\n",
+    "    similarity[feature + 'B' + suffix] = cell[feature] and not neighbour[feature]\n",
+    "    return similarity\n",
+    "    \n",
+    "def extractSimilarityFeatures(cell, neighbour, suffix):\n",
+    "    similarityFeatures = {\n",
+    "        **getSimilarity('isMerged', cell, neighbour, suffix),\n",
+    "        **getSimilarity('isCenterAligned', cell, neighbour, suffix),\n",
+    "        **getSimilarity('isTHOrInTHead', cell, neighbour, suffix),\n",
+    "        **getSimilarity('isBold', cell, neighbour, suffix),\n",
+    "        **getSimilarity('isItalic', cell, neighbour, suffix),\n",
+    "        **getSimilarity('isUnderlined', cell, neighbour, suffix),\n",
+    "        **getSimilarity('isEmpty', cell, neighbour, suffix),\n",
+    "        **getSimilarity('isText', cell, neighbour, suffix),\n",
+    "        **getSimilarity('isNumeric', cell, neighbour, suffix),\n",
+    "        **getSimilarity('isDate', cell, neighbour, suffix),\n",
+    "        **getSimilarity('isShortText', cell, neighbour, suffix),\n",
+    "        **getSimilarity('isLongText', cell, neighbour, suffix),\n",
+    "        **getSimilarity('isTotal', cell, neighbour, suffix)        \n",
+    "    }\n",
+    "    return similarityFeatures    \n",
+    "    \n",
+    "def addSimilarityFeatures(normalizedFeatureTable):\n",
+    "    nftWithSimilarity = []\n",
+    "    numRows = len(normalizedFeatureTable)\n",
+    "    for rowIndex, row in enumerate(normalizedFeatureTable):\n",
+    "        nftWithSimilarity.append([])\n",
+    "        for cellIndex, cell in enumerate(row):\n",
+    "            features = cell\n",
+    "            if rowIndex > 0:\n",
+    "                features = {\n",
+    "                    **extractSimilarityFeatures(cell, normalizedFeatureTable[rowIndex - 1][cellIndex], 'u'), \n",
+    "                    **features\n",
+    "                    }\n",
+    "            if rowIndex < numRows - 1:\n",
+    "                features = {\n",
+    "                    **extractSimilarityFeatures(cell, normalizedFeatureTable[rowIndex + 1][cellIndex], 'l'),\n",
+    "                    **features\n",
+    "                }\n",
+    "            intCellFeatures = mapDictBoolValuesToInt(features)\n",
+    "            nftWithSimilarity[-1].append(intCellFeatures)\n",
+    "    return nftWithSimilarity\n",
+    "    \n",
+    "def cleanOfEmptyCells(table):\n",
+    "    lastEmptyCellIndex = len(table[0])\n",
+    "    for row in table:\n",
+    "        for cellIndex,cell in enumerate(row):\n",
+    "            if cell == 'empty cell':\n",
+    "                lastEmptyCellIndex = min(lastEmptyCellIndex, cellIndex)\n",
+    "    newTable =[]\n",
+    "    for row in table:\n",
+    "        newTable.append(row[:lastEmptyCellIndex])\n",
+    "    return newTable\n",
+    "\n",
+    "def normalizedFeatureTable(table):\n",
```

Comment:
Personally, I'm not a big fan of abbreviated varibale names. I would go for normalizedFeatureTable instead of nft.
This method is also kinda long. Can we split it down by moving code into their own methods and giving them more meaning by the method name? That would also help understand what the code does here. Potentially where ever we have nested loops that could indicate that we can make use of splitting code into methods.